File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/transaction/ServiceA.java
package org.babyfish.jimmer.sql.transaction;

import org.babyfish.jimmer.sql.JSqlClient;

@TargetAnnotation(Component.class)
@Tx(Propagation.MANDATORY)
public class ServiceA {

    protected final JSqlClient sqlClient;

    public ServiceA(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    public int a() {
        return 0;
    }

    @Tx(Propagation.REQUIRES_NEW)
    void b() {}
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/transaction/ServiceA.kt
package org.babyfish.jimmer.sql.kt.transaction

import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.transaction.Propagation
import org.babyfish.jimmer.sql.transaction.TargetAnnotation
import org.babyfish.jimmer.sql.transaction.Tx

@TargetAnnotation(Component::class)
@Tx(Propagation.MANDATORY)
open class ServiceA(
    protected val sqlClient: KSqlClient
) {

    open fun a() {}

    @Tx(Propagation.REQUIRES_NEW)
    internal open fun b() {}
}